lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1538 @@
+    } catch {
+        sleep 1
+        Start-ExternalCommand { git clean -x -f -d -q } -ErrorMessage "Failed to git clean"
+    }
+
+    echo "Before doing git checkout:"
+    echo "Git branch output:"
+    Start-ExternalCommand { git branch } -ErrorMessage "Failed to show git branch."
+    echo "Git log output:"
+    Start-ExternalCommand { git log -10 --pretty=format:"%h - %an, %ae, %ar : %s" } `
+        -ErrorMessage "Failed to show git log."
+
+    $ret = echo "$ZuulRef" | Where-Object { $_ -match "^refs/tags/" }
+    if ($ret) {
+        Start-ExternalCommand { git fetch --tags "$ZuulUrl/$ZuulProject" } `
+            -ErrorMessage "Failed to fetch tags from: $ZuulUrl/$ZuulProject"
+        Start-ExternalCommand { git checkout $ZuulRef } `
+            -ErrorMessage "Failed to fetch tags to: $ZuulRef"
+        Start-ExternalCommand { git reset --hard $ZuulRef } `
+            -ErrorMessage "Failed to hard reset to: $ZuulRef"
+    } elseif (!$ZuulNewrev) {
+        Start-ExternalCommand { git fetch "$ZuulUrl/$ZuulProject" $ZuulRef } `
+            -ErrorMessage "Failed to fetch: $ZuulUrl/$ZuulProject $ZuulRef"
+        Start-ExternalCommand { git checkout FETCH_HEAD } `
+            -ErrorMessage "Failed to checkout FETCH_HEAD"
+        Start-ExternalCommand { git reset --hard FETCH_HEAD } `
+            -ErrorMessage "Failed to hard reset FETCH_HEAD"
+    } else {
+        Start-ExternalCommand { git checkout $ZuulNewrev } `
+            -ErrorMessage "Failed to checkout $ZuulNewrev"
+        Start-ExternalCommand { git reset --hard $ZuulNewrev } `
+            -ErrorMessage "Failed to hard reset $ZuulNewrev"
+    }
+
+    try {
+        Start-ExternalCommand { git clean -x -f -d -q } -ErrorMessage "Failed to git clean"
+    } catch {
+        sleep 1
+        Start-ExternalCommand { git clean -x -f -d -q } -ErrorMessage "Failed to git clean"
+    }
+
+    if (Test-Path .gitmodules) {
+        Start-ExternalCommand { git submodule init } -ErrorMessage "Failed to init submodule"
+        Start-ExternalCommand { git submodule sync } -ErrorMessage "Failed to sync submodule"
+        Start-ExternalCommand { git submodule update --init } -ErrorMessage "Failed to update submodule"
+    }
+
+    echo "Final result:"
+    echo "Git branch output:"
+    Start-ExternalCommand { git branch } -ErrorMessage "Failed to show git branch."
+    echo "Git log output:"
+    Start-ExternalCommand { git log -10 --pretty=format:"%h - %an, %ae, %ar : %s" } `
+        -ErrorMessage "Failed to show git log."
+
+    popd
+}
+
+
+function Install-Projects ($buildFor) {
+    # Shared for all projects
+    $projectDir = $buildFor -replace 'openstack/'
+    Write-JujuLog "Installing $projectDir"
+    $openstackBuild = Join-Path $BUILD_DIR "openstack"
+    Start-ExecuteWithRetry {
+        Install-OpenStackProjectFromRepo "$openstackBuild\$projectDir"
+    }
+    
+    # Individual projects
+    # Nova
+    if ($buildFor -eq "openstack/nova") {
```

Comment:
You have two big if blocks here. I would separate the contents of thos if blocks into separate functions, and just call those functions in this one.

Something like ```Invoke-BuildNova``` and ```Invoke-BuildNeutron```